nx/lib/nx.ex

```diff
@@ -1120,6 +1120,91 @@ defmodule Nx do
     backend.eye(%T{type: type, shape: shape, names: names}, backend_options)
   end
 
+  @doc """
+  Extract a diagonal or create a diagonal tensor.
+
+  Extracts a diagonal if given a 2d tensor, or constructs a tensor given a 1d one.
+
+  ## Examples
+
+  When given a 1d tensor:
+
+      iex> Nx.iota({4}) |> Nx.diag
+      #Nx.Tensor<
+        s64[4][4]
+        [
+          [0, 0, 0, 0],
+          [0, 1, 0, 0],
+          [0, 0, 2, 0],
+          [0, 0, 0, 3]
+        ]
+      >
+
+  Given a 2d tensor:
+
+      iex> Nx.iota({3, 3}) |> Nx.diag()
+      #Nx.Tensor<
+        s64[3]
+        [0, 4, 8]
+      >
+
+  And if given a 2d tensor along with an offset:
+
+      iex> Nx.iota({3, 3}) |> Nx.diag(offset: 1)
+      #Nx.Tensor<
+        s64[2]
+        [1, 5]
+      >
+
+      iex> Nx.iota({3, 3}) |> Nx.diag(offset: -1)
+      #Nx.Tensor<
+        s64[2]
+        [3, 7]
+      >
+
+  ## Options
+
+    * `:offset` - offset used for extracting the diagonal.
+      Use offset > 0 for diagonals above the main diagonal,
+      and offset < 0 for diagonals below the main diagonal.
+
+  """
+  @doc type: :creation
+  def diag(tensor, opts \\ [])
+
+  def diag(tensor = %Nx.Tensor{shape: {len, breadth}}, opts) do
+    opts = keyword!(opts, [:offset])
+
+    offset = opts[:offset] || 0
+
+    indices =
+      case offset do
+        i when i >= 0 ->
+          Enum.zip_with(0..(len - 1), i..(breadth - 1), fn x, y -> [x, y] end)
+
+        i when i < 0 ->
+          Enum.zip_with(-i..(len - 1), 0..(breadth - 1), fn x, y -> [x, y] end)
+      end
+
+    indices =
+      indices
+      |> Nx.tensor()
+
+    Nx.gather(tensor, indices)
+  end
+
+  def diag(tensor = %Nx.Tensor{shape: {len}}, _opts) do
+    identity =
+      {len, len}
+      |> Nx.eye()
+
+    Nx.multiply(tensor, identity)
+  end
+
+  def diag(tensor, _opts) do
+    raise ArgumentError, "diag/2 expects tensor of rank 1 or 2, got: #{inspect(tensor)}"
+  end
```

Comment:
we normally do shape checks with `Nx.Shape.<function_name>`. Perhaps you could move the separate clauses to a `defp extract_diag` and `defp build_diag` or something like that to keep with the pattern

Reply:
@polvalente,

Could you specify what you mean? I don't quite understand.. Is there a `Nx.Shape.shape!` function that you expect me to use, or am I supposed to do something else...

Reply:
Instead of hard-matching on the tensor shapes you would delegate the check for 2D vs 1D (and this raise) do a `Nx.Shape.diag` function. Don't worry about this, though. We can change this later if need comes :)